packages/node-provider/src/node-provider.ts

```diff
@@ -1,7 +1,66 @@
-import { INodeProvider, NodeMessage } from "./types";
+import EventEmitter from "eventemitter3";
+
+import {
+  INodeProvider,
+  NodeMessage,
+  NodeMessageReceivedCallback
+} from "./types";
 
 export default class NodeProvider implements INodeProvider {
-  public onMessage(callback: (message: NodeMessage) => void) {}
+  private isConnected: boolean;
+  private eventEmitter: EventEmitter;
+  private messagePort?: MessagePort;
+
+  constructor() {
+    this.isConnected = false;
+    this.eventEmitter = new EventEmitter();
+  }
+
+  public onMessage(callback: NodeMessageReceivedCallback) {
+    this.eventEmitter.on("message", callback);
+  }
+
+  public sendMessage(message: NodeMessage) {
+    if (!this.isConnected || !this.messagePort) {
+      throw new Error(
+        "It's not possible to use postMessage() before the NodeProvider is connected. Call the connect() method first."
+      );
+    }
+
+    this.messagePort.postMessage(message);
+  }
+
+  public async connect(): Promise<NodeProvider> {
+    if (this.isConnected) {
+      console.warn("NodeProvider is already connected.");
+      return Promise.resolve(this);
+    }
+
+    return new Promise<NodeProvider>(this.getMessagePort.bind(this));
+  }
+
+  private getMessagePort(resolve, reject) {
```

Comment:
I think you should inline this method. It's confusing that it's separate from the method in which the Promise is created.